From d8123a0ec23bb4c7c82b35227098958ee88652d6 Mon Sep 17 00:00:00 2001
From: Lei wang <leiwan@codeaurora.org>
Date: Thu, 10 May 2018 17:53:58 +0800
Subject: [PATCH] kernel: early-init: support early smack rules

For early userpace, add some basic rules.
when system rules loaded later, will switch to
system rules

Change-Id: Ie04f41c543aa6312cae5b400be6c2f3a76f50d62
Signed-off-by: Lei wang <leiwan@codeaurora.org>
---
 security/smack/smack_access.c | 9 ++++++++-
 1 file changed, 8 insertions(+), 1 deletion(-)

diff --git a/security/smack/smack_access.c b/security/smack/smack_access.c
index a283f9e..ca66d00 100644
--- a/security/smack/smack_access.c
+++ b/security/smack/smack_access.c
@@ -87,6 +87,8 @@ int log_policy = SMACK_AUDIT_DENIED;
  *
  * Allowing write access implies allowing locking.
  */
+static bool is_rule_empty = true;
+
 int smk_access_entry(char *subject_label, char *object_label,
 			struct list_head *rule_list)
 {
@@ -94,13 +96,18 @@ int smk_access_entry(char *subject_label, char *object_label,
 	struct smack_rule *srp;
 
 	list_for_each_entry_rcu(srp, rule_list, list) {
+		is_rule_empty = false;
 		if (srp->smk_object->smk_known == object_label &&
 		    srp->smk_subject->smk_known == subject_label) {
 			may = srp->smk_access;
 			break;
 		}
 	}
-
+	/*
+	 * Support early userspace smack checking without rules loaded
+	 */
+	if (is_rule_empty && (0 == strcmp(subject_label, "System")))
+		may = MAY_WRITE | MAY_READ | MAY_EXEC;
 	/*
 	 * MAY_WRITE implies MAY_LOCK.
 	 */
-- 
1.9.1

